parser/ParseIndexProperties.hpp

```diff
@@ -0,0 +1,412 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *     University of Wisconsin—Madison.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#ifndef QUICKSTEP_PARSER_PARSE_INDEX_PROPERTIES_HPP_
+#define QUICKSTEP_PARSER_PARSE_INDEX_PROPERTIES_HPP_
+
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <vector>
+#include <unordered_map>
+
+#include "parser/ParseKeyValue.hpp"
+#include "parser/ParseString.hpp"
+#include "parser/ParseTreeNode.hpp"
+#include "storage/StorageBlockLayout.pb.h"
+#include "utility/Macros.hpp"
+#include "utility/PtrList.hpp"
+#include "utility/StringUtil.hpp"
+
+#include "glog/logging.h"
+
+namespace quickstep {
+/** \addtogroup Parser
+ *  @{
+ */
+
+/**
+ * @brief Base class for different types of indices and their associated properties
+ */
+class IndexProperties {
+ public:
+  typedef std::unordered_map<const char*, ParseKeyValue::KeyValueType> valid_property_map_type;
+
+  // An enum specifying the reason for an invalid index description, if any.
+  enum class InvalidIndexType {
+    kNone = 0,
+    kUnimplemented,
+    kDuplicateKey,
+    kInvalidKey,
+    kSizeIsFloat,
+    kSizeIsNegative,
+    kNumHashesIsFloat,
+    kNumHashesIsNegative,
+    kProjectedCountIsFloat,
+    kProjectedCountIsNegative
+  };
+
+  /**
+   * @brief Constructor.
+   *
```

Comment:
Add comments about `@note The constructor takes ownership of \c index_sub_block_description.`